### DISCUSSION TITLE
Initial configuration script within user data when creating nodes

### DISCUSSION THREAD

**Discussion by ghost** [bottlerocket-os/bottlerocket#3191] [Q&A]
Good morning,

We are looking for a way to standardize configuration across all our EKS clusters and use this script to pull configurations such as container registry credentials, mirroring configuration, SSH keys for automation processes through Ansible, the initial "version-lock" that should apply to the node to match the current upgrade state of all other nodes in the cluster, etc... from a git repository.

What are the options to execute a shell script as part of the user data injected by AWS from the launch template? Could you please advise on a recommended procedure or best practices?

Thank you in advance,
ajdelaguila

**Comment by armenr** [comment]
That's going to depend on the way you're deploying (and scaling) your clusters. If it's with terraform, you need to ensure you have the right BottleRocket-specific UserData passed in.

If you're using Karpenter for scaling worker nodes on EKS, you'll also want to make sure you have the same UserData in your NodeTemplate(s). 

We've been migrating more and more of our new infrastructure architecture to BottleRocket from AL2. One of the biggest differences is UserData. In AL2, you can pass a script to UserData and have it run when the machine boots.

In BottleRocket, a good way to think about things is that *everything* is jailed inside a container. This is by design...for the sake of security, but also helps to make the system more reason-about-able. 

https://github.com/bottlerocket-os/bottlerocket#bootstrap-containers-settings

You're going to want to read about bootstrap containers. The shell scripts or Ansible PlayBooks you wish to run as part of your UserData bootstrap process will have to be packaged up as container images, with the correct ENTRYPOINT and/or CMD in the Dockerfile. 

You'll then declare and configure a bootstrap container for BottleRocket which pulls and runs that container on the BottleRocket host.

Bootstrap containers are priveleged, and should provide the filesystem-level access you need to the host in order to set everything up.

"Converting" your brain to think about things in a "BottleRocket-friendly way" takes a little time. The whole OS exposes an API for itself (which takes some time getting used to), and it essentially requires that you use Docker containers for any host-related interactions with the underlying system, similar to the way you'd think of using old-school chroot-ed jails on a traditional server in order to do specific things on that host, without exposing the whole host to your specifically jailed scripts or executables. 

I defer to the BottleRocket maintainers in case there's something I've misrepresented, or there's something that I - myself - have misunderstood.

**Comment by ghost** [comment]
@armenr Thank you very much for sharing your thoughts. It matches what I have understood from reading the documentation as well and I am pretty comfortable with that approach.

I bet that looking at the default bootstrap container image, its entry point and its definition would give an idea of how to inject our commands before the "official" behaviour kicks in. If we implement the integration this way, we can still base our container image in the official bootstrap, and enjoy our automation pieces, but continue with the official configuration method to not miss any critical configuration made by the default one.

**Comment by arnaldo2792** [comment]
Hi @ajdelaguila-rbro, there isn't any default / official bootstrap container image, although there are a few examples that different users have posted across a few discussions. Bottlerocket doesn't define any default bootstrapping through bootstrap containers, they are a feature to allow users set up their hosts for cases that aren't solved through the API (e.g., mount an ephemeral disk to either `/var/lib/docker` or `/var/lib/containerd`). So if you already have a container with the configuration that you want to apply, you may have to tweak it a little bit to set up the configurations that you need in a Bottlerocket host.

Keep in mind that, even though bootstrap containers have some privilege, they don't run with full privilege and the things they are allowed to do are limited. Please refer to the documentation to know what [capabilities(7)] they were granted.

@armenr, that was an excellent answer! Although I want to make a small clarification. Both host and bootstrap containers aren't Docker containers. They don't run with Docker as the container runtime, rather Bottlerocket ships with a small container client called `host-ctr` which communicates directly with a separate instance of containerd.

[capabilities(7)]: https://man7.org/linux/man-pages/man7/capabilities.7.html

**Comment by ghost** [comment]
@arnaldo2792 Could you please lend us a hand with the bootstrap container?

We have built it to use as our bootstrap container, the entry point of the image is set to run a shell script that, right now, just echoes a message to the standard output.

```shell
#!/bin/bash

echo "Initializing Bottlerocket Bootstrap..."
```

User data for the node is configured as:
```
user_data = base64encode(templatefile("${path.module}/bottlerocket_user_data.tpl", {
    cluster_name         = aws_eks_cluster.eks-cluster.name
    cluster_endpoint     = aws_eks_cluster.eks-cluster.endpoint
    cluster_auth_base64  = aws_eks_cluster.eks-cluster.certificate_authority[0].data
    bootstrap_extra_args = <<-EOT
        [settings.pki.my-cert-bundle]
        data = "********"
        trusted = true
        [settings.bootstrap-containers.bootstrap]
        source = "artifactory.tech.rbro.rbg.cc/docker-secure/devops/bottlerocket-bootstrap:latest"
        mode = "always"
        essential = false
        [[settings.container-registry.credentials]]
        registry = "artifactory.tech.rbro.rbg.cc"
        username = "********"
        password = "********"
      EOT
  }))
}
```

If we set "essential" to true, the node never makes it up nor joins the EKS cluster. Following some other threads here in GitHub, we have set it up to "false" and gathered the logs from inside the node.

In journalctl.log, among other initialization steps, we can see this:
```
 26 16:13:31 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:31.184835169Z" level=info msg=serving... address=/run/host-containerd/containerd.sock
Jun 26 16:13:31 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Started containerd runtime for host containers.
Jun 26 16:13:31 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:31.185274626Z" level=info msg="containerd successfully booted in 0.038338s"
Jun 26 16:13:31 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting bootstrap container bootstrap...
Jun 26 16:13:31 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:31Z" level=info msg="Image does not exist, proceeding to pull image from source." ref="artifactory.tech.rbro.rbg.cc/docker-secure/devops/bottlerocket-bootstrap:latest"
Jun 26 16:13:33 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:33Z" level=info msg="pulled image successfully" img="artifactory.tech.rbro.rbg.cc/docker-secure/devops/bottlerocket-bootstrap:latest"
Jun 26 16:13:33 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:33Z" level=info msg="unpacking image..." img="artifactory.tech.rbro.rbg.cc/docker-secure/devops/bottlerocket-bootstrap:latest"
Jun 26 16:13:36 ip-10-224-32-44.eu-central-1.compute.internal chronyd[1135]: 2023-06-26T16:13:36Z Selected source 169.254.169.123
Jun 26 16:13:38 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: var-lib-host\x2dcontainerd-tmpmounts-containerd\x2dmount4095688669.mount: Deactivated successfully.
Jun 26 16:13:41 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:41Z" level=info msg="Container does not exist, proceeding to create it" ctr-id=boot.bootstrap
Jun 26 16:13:41 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: tmp-containerd\x2dmount3838552900.mount: Deactivated successfully.
Jun 26 16:13:41 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:41Z" level=info msg="container task does not exist, proceeding to create it" container-id=boot.bootstrap
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:42.003620809Z" level=info msg="loading plugin \"io.containerd.event.v1.publisher\"..." runtime=io.containerd.runc.v2 type=io.containerd.event.v1
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:42.003774731Z" level=info msg="loading plugin \"io.containerd.internal.v1.shutdown\"..." runtime=io.containerd.runc.v2 type=io.containerd.internal.v1
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:42.003823612Z" level=info msg="loading plugin \"io.containerd.ttrpc.v1.task\"..." runtime=io.containerd.runc.v2 type=io.containerd.ttrpc.v1
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:42.004039116Z" level=info msg="starting signal loop" namespace=default path=/run/host-containerd/io.containerd.runtime.v2.task/default/boot.bootstrap pid=1167 runtime=io.containerd.runc.v2
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: Initializing Bottlerocket Bootstrap...
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:42Z" level=info msg="successfully started container task"
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal host-ctr[1146]: time="2023-06-26T16:13:42Z" level=info msg="container task exited" code=0
Jun 26 16:13:42 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: run-host\x2dcontainerd-io.containerd.runtime.v2.task-default-boot.bootstrap-rootfs.mount: Deactivated successfully.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:45.549821964Z" level=info msg="shim disconnected" id=boot.bootstrap
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:45.549919816Z" level=warning msg="cleaning up after shim disconnected" id=boot.bootstrap namespace=default
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:45.549939156Z" level=info msg="cleaning up dead shim"
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal containerd[1136]: time="2023-06-26T16:13:45.562331971Z" level=warning msg="cleanup warnings time=\"2023-06-26T16:13:45Z\" level=info msg=\"starting signal loop\" namespace=default pid=1207 runtime=io.containerd.runc.v2\n"
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal bootstrap-containers[1220]: 16:13:45 [INFO] bootstrap-containers started
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal bootstrap-containers[1220]: 16:13:45 [INFO] Mode for 'bootstrap' is 'always'
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Finished bootstrap container bootstrap.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Reached target Bottlerocket final configuration complete.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Reboot system if needed for configuration changes to take effect...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal prairiedog[1223]: 16:13:45 [INFO] No reboot required
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: reboot-if-required.service: Deactivated successfully.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Finished Reboot system if needed for configuration changes to take effect.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Isolates multi-user.target...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemctl[1227]: Removed /etc/systemd/system/default.target.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemctl[1227]: Created symlink /etc/systemd/system/default.target → /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/multi-user.target.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Reloading.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Created slice Kubernetes and container runtime slice.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Created slice Slice /system/host-containers.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Reached target Login Prompts.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Reached target Remote File Systems.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting containerd container runtime...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Started Host container: control.
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Load crash kernel was skipped because of a failed condition check (ConditionKernelCommandLine=crashkernel).
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Load Kernel Module ip_vs_rr...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Load Kernel Module ip_vs_sh...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Load Kernel Module ip_vs_wrr...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Reboot system if needed for configuration changes to take effect...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Starting Send boot success...
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal systemd[1]: Platform Persistent Storage Archival was skipped because of a failed condition check (ConditionDirectoryNotEmpty=/sys/fs/pstore).
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal shibaken[1252]: 16:13:45 [INFO] shibaken started
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal shibaken[1252]: autoscaling.should-wait value is false
Jun 26 16:13:45 ip-10-224-32-44.eu-central-1.compute.internal shibaken[1252]: Marker file path is /var/lib/bottlerocket/warm-pool-wait.ran
```

We understand from the log that the bootstrap container has been loaded and has completed its entry point also successfully.

What are we doing wrong?

Thanks in advance!
Antonio Jesus

**Comment by arnaldo2792** [comment]
Let me dig into this, and try to reproduce

**Comment by arnaldo2792** [comment]
Hey @ajdelaguila-rbro, I wasn't able to replicate what you are experiencing. I created a container with this Dockerfile:

```Dockerfile
FROM alpine

# Just for debugging purpuses
RUN apk add coreutils

ENTRYPOINT ["echo", "HI, I'm bootstraping!"]
```

Then, I updated my API settings to configure a new bootstrap container:

```bash
{
  "settings": {
    "bootstrap-containers": {
      "bootstrap": {
        "essential": true,
        "mode": "always",
        "source": "<my-image>/bootstrap:latest"
      }
    }
  }
}

```

My instance came back up after reboot, and the bootstrap container was executed:

```bash
# From the admin container
[root@admin]# sheltie journalctl -u bootstrap-containers@bootstrap.service
Jun 28 20:47:19 <> host-ctr[1764]: time="2023-06-28T20:47:19Z" level=info msg="successfully started container task"
Jun 28 20:47:19 <> host-ctr[1764]: HI, I'm bootstraping!
Jun 28 20:47:19 <> host-ctr[1764]: time="2023-06-28T20:47:19Z" level=info msg="container task exited" code=0
Jun 28 20:47:20 <> bootstrap-containers[1833]: 20:47:20 [INFO] bootstrap-containers started
Jun 28 20:47:20 <> bootstrap-containers[1833]: 20:47:20 [INFO] Mode for 'bootstrap' is 'always'
Jun 28 20:47:20 <> systemd[1]: Finished bootstrap container bootstrap.
```

Could you please share how you are creating your bootstrap container? Also, if you can get the logs of an instance that doesn't join a cluster after you apply your changes will help! :+1:

**Comment by ghost** [comment]
Good afternoon @arnaldo2792,

Here is my Dockerfile:
```
FROM artifactory.tech.rbro.rbg.cc/registry.redhat.io/rhel7/rhel-tools:latest

LABEL name="bottlerocket-bootstrap" \
  version="9.x" \
  maintainer="Cloud Team <cloud@raiffeisen.ro>" \
  architecture="x86_64" \
  io.k8s.display-name="RBRO :: Bottlerocket Bootstrap" \
  io.k8s.description="Container image to customize a Bottlerocket host part of an EKS cluster in RBRO." \
  io.openshift.tags="eks,bottlerocket,bootstrap,customize"

USER root

COPY cacerts/* /etc/pki/ca-trust/source/anchors/
RUN chmod 644 /etc/pki/ca-trust/source/anchors/* && update-ca-trust

ARG RBRO_CONTAINER_USERNAME="cloud-user"
ARG RBRO_CONTAINER_USER_GROUP_IDS="1001:0"
ARG RBRO_CONTAINER_HOME="/home/$RBRO_CONTAINER_USERNAME"

# Copy rpm packages to the tmp folder
COPY rpms/* /tmp/

# Install dependencies
# RUN yum-config-manager --disable * && \
#   yum-config-manager --enable rhel-server-rhscl-7-rpms --enable RBRO_EPEL_epel_rhel7 && \
#   INSTALL_PKGS="jq" && \
#   yum install  --nogpgcheck -y $INSTALL_PKGS && \
#   yum clean all -y && \
RUN yum localinstall /tmp/*.rpm -y && \
  rm -fr /tmp/*.rpm && \
  chmod g+rw /etc/profile.d/sh.local && \
  echo '$RBRO_CONTAINER_USERNAME:x:$RBRO_CONTAINER_USER_GROUP_IDS::$RBRO_CONTAINER_HOME:/bin/bash\n' >> /etc/passwd && \
  mkdir -p $RBRO_CONTAINER_HOME

COPY bottlerocket-settings.json $RBRO_CONTAINER_HOME/
COPY configure.sh $RBRO_CONTAINER_HOME/

RUN chown -R $RBRO_CONTAINER_USER_GROUP_IDS $RBRO_CONTAINER_HOME && \
  chmod -R g+rw $RBRO_CONTAINER_HOME && \
  chmod g+rwx $RBRO_CONTAINER_HOME/configure.sh

USER $RBRO_CONTAINER_USER_GROUP_IDS

ENTRYPOINT ["sh", "/home/cloud-user/configure.sh"]
```

and this is the content of the configure.sh script
```shell
#!/bin/bash

echo "Initializing Bottlerocket Bootstrap..."

# Load content of the settings file in environment variable
# BOTTLEROCKET_JSON_SETTINGS=$(jq -cr . /etc/bottlerocket-settings.json)
# Send configuration to BottleRocket through its cli
# apiclient set --json '$BOTTLEROCKET_JSON_SETTINGS'

```

I've tried running this image using just docker and from Kubernetes and in both cases it traces the message inside the script configre.sh

We're now trying to check again if we can gain access through SSH when the node is in that state; the previous try we did, we couldn't connect.

Is there something in the Dockerfile that catches your attention? Is there something you would recommend testing/trying?

Thank you in advance,
Antonio Jesus

**Comment by ghost** [comment]
@arnaldo2792,

Today we tried again and it worked right away, we got the message in the log and the node joined the cluster with no issue.

We believe that the behaviour is due to some cache timeout that prevents the image from been always pulled during boot time. If it's present in the node, it never checks if the SHA code of the image in the remote repository matches the one pulled locally, and it does not download it again. 

Last Monday, when we ran this procedure, we realized that our bootstrap image had an issue during its initialization and we built a second version of the image (the one that today worked right away). Today, after getting this working we have updated our bootstrap image to perform actual configuration changes and they have not been executed in any of the nodes. However, they have been performed in a new node added to the cluster.

Since we are talking about a bootstrap image that needs to be configured when the cluster is first created, we are pointing to an elastic tag (latest) that we move as we release newer versions of the image. We have checked that if we modify the node group to change the source of the bootstrap image to point to a different version it recreates all nodes that belong to the node group causing a major outage in the cluster and we want to avoid that situation.

Is there any configuration value we can set so the image is always pulled similarly to the flag in Kubernetes deployments? If not, what approach would you recommend?

Thank you again for your time and support,
Antonio Jesus

**Reply by arnaldo2792** [reply]
Hey @ajdelaguila-rbro, we do have a flag to use a cached image but for bootstrap containers we always pull. I also confirmed that after I pushed twice to `latest`, the image was re-pulled on reboot:

```
bash-5.1# journalctl -u bootstrap-containers@bootstrap.service | grep HI
Jun 29 23:33:42 <> host-ctr[1821]: HI, I'm bootstraping!
Jun 29 23:35:56 <> host-ctr[1769]: HI, I'm bootstraping again!
```

Is there anything special about the registry you are using? Does it cache an image somehow?

**Comment by ghost** [comment]
Good morning @arnaldo2792,

We use Artifactory as a container registry and we have confirmed it always serves the actual image behind the tag with no caching.

I wonder if this behaviour might be related to the fact that we are testing this implementation on version 1.13.0, we'll give this a try on 1.14.1 as well, just to confirm.

On the other hand, when running "apiclient" commands from the bootstrap container we receive "Permission denied.". Is there any particular user id or group id we need to use in our bootstrap image? Following standard good practices when creating containers we are not using the "root" user but user 1001; however, for this particular use case, should we run the commands using the root user of the container?

Have a wonderful day,
Antonio Jesus.

**Comment by ghost** [comment]
@arnaldo2792 

I can confirm that using the root user within the bootstrap container fixes the permissions issue and now commands go through as expected.

Regarding the cache... there might be some corner cases when the bootstrap image fails to run or OS v1.13.0 or something similar because today we get the new container pulled every time.

I think that, at this point, we are good to close this conversation, everything is working as intended.

@arnaldo2792, We cannot thank you enough for all the time you've dedicated and the help provided, 

Have a great weekend ahead,
Antonio Jesus

**Comment by arnaldo2792** [comment]
>  Following standard good practices

@ajdelaguila-rbro, I'm glad you folks got your setup working! :tada: , I'm sorry I missed your update about the user ID, that's unexpected, let me try to confirm this since in host containers the user isn't UID 0, and they can execute `apiclient`.

```bash
[ssm-user@control]$ id -u
1000
[ssm-user@control]$ apiclient get os
[ssm-user@control]$ id -u
1000
[ssm-user@control]$ apiclient get settings.host-containers.control
{
  "settings": {
    "host-containers": {
      "control": {
        "enabled": true,
        "source": "328549459982.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-control:v0.7.2",
        "superpowered": false
      }
    }
  }
}
```

Regarding the 1.13.0 version, there were several patches to that particular version, would it be possible for you folks to move to the latest version of that release? Or even better, to 1.14.X?
